torchgeo/datasets/skyscript.py

```diff
@@ -0,0 +1,187 @@
+
+    image_dirs = tuple(f'images{i}' for i in range(2, 8))
+    image_md5s = (
+        'fbfb5f7aa1731f4106fc3ffbd608100a',
+        'ad4fd9fdb9622d1ea360210cb222f2bd',
+        'aeeb41e830304c74b14b5ffc1fc8e8c3',
+        '02ee7e0e59f9ac1c87b678a155e1f1df',
+        '350475f1e7fb996152fa16db891b4142',
+        '5e2fbf3e9262b36e30b458ec9a1df625',
+    )
+
+    #: Can be modified in subclasses to change train/val/test split
+    caption_files: ClassVar[dict[str, str]] = {
+        'train': 'SkyScript_train_top30pct_filtered_by_CLIP_openai.csv',
+        'val': 'SkyScript_val_5K_filtered_by_CLIP_openai.csv',
+        'test': 'SkyScript_test_30K_filtered_by_CLIP_openai.csv',
+    }
+    caption_md5s: ClassVar[dict[str, str]] = {
+        'train': '05b362e43a852667b5374c9a5ae53f8e',
+        'val': 'c8d278fd29b754361989d5e7a6608f69',
+        'test': '0135d9b49ce6751360912a4353e809dc',
+    }
+
+    def __init__(
+        self,
+        root: Path = 'data',
+        split: str = 'train',
+        transforms: Callable[[dict[str, Tensor]], dict[str, Tensor]] | None = None,
+        download: bool = False,
+        checksum: bool = False,
+    ) -> None:
+        """Initialize a new SkyScript instance.
+
+        Args:
+            root: Root directory where dataset can be found.
+            split: One of 'train', 'val', 'test'.
+            transforms: A function/transform that takes input sample and its target as
+                entry and returns a transformed version.
+            download: If True, download dataset and store it in the root directory.
+            checksum: If True, check the MD5 of the downloaded files (may be slow).
+
+        Raises:
+            AssertionError: If *split* is invalid.
+            DatasetNotFoundError: If dataset is not found and *download* is False.
+        """
+        assert split in self.caption_files
+
+        self.root = root
+        self.split = split
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+
+        self._verify()
+
+        self.captions = pd.read_csv(os.path.join(self.root, self.caption_files[split]))
+
+    def __len__(self) -> int:
+        """Return the number of images in the dataset.
+
+        Returns:
+            Length of the dataset.
+        """
+        return len(self.captions)
+
+    def __getitem__(self, index: int) -> dict[str, Any]:
+        """Return an index within the dataset.
+
+        Args:
+            index: Index to return.
+
+        Returns:
+            A dict containing image and caption at index.
+        """
+        filepath, title = self.captions.iloc[index][:2]
+
+        with Image.open(os.path.join(self.root, filepath)) as img:
+            array = np.array(img, dtype=np.float32)
+            array = rearrange(array, 'h w c -> c h w')
+            image = torch.from_numpy(array)
+
+        sample = {'image': image, 'text': title}
```

Comment:
Since this is our first text dataset, we have a couple of things to decide here:

* sample key: text? caption? title?
* value dtype: str? Tensor?

Also, all images are completely different sizes (ranging from 100 px to 1000 px). We should decide whether we want to crop/resize in the dataset or force the user to add a transform to do this.

Reply:
I think "text" is the most neutral, since it could be a caption, a title, a description, or a prompt etc. In the end people need to apply a text embedding so str for now is fine, since the options for different text embeddings is large.

And I do like giving the crop/resize option at the datamodule level, to have one implementation that is "low effort" to get started.

Reply:
I'm actually leaning towards "caption" now. For example, imagine a geo-embedded dataset with text as input and output. We need some way to distinguish between the two. But let me ask a couple NLP people and see if there is a standard w.r.t. augmentation libraries.

Reply:
Torchvision uses "caption" for COCO, which seems to be the gold standard dataset for image captioning: https://github.com/pytorch/vision/blob/main/torchvision/datasets/coco.py